src/tools/compiletest/src/header.rs

```diff
@@ -1113,34 +1113,45 @@ fn parse_normalize_rule(header: &str) -> Option<(String, String)> {
     Some((regex, replacement))
 }
 
-pub fn extract_llvm_version(version: &str) -> Option<u32> {
-    let pat = |c: char| !c.is_ascii_digit() && c != '.';
-    let version_without_suffix = match version.find(pat) {
-        Some(pos) => &version[..pos],
+/// Given an llvm version string that looks like `1.2.3-rc1`, extract key version info into a
+/// [`LlvmVersion`].
+///
+/// Currently panics if the input string is malformed, though we really should not use panic as an
+/// error handling strategy.
+///
+/// FIXME(jieyouxu): improve error handling
```

Comment:
Did you consider using `semver`?

Reply:
I did not immediately, since it wasn't in the dependencies for compiletest specifically, although we certainly could as I imagine `semver` is already in the workspace dep anyway.

Reply:
I tried this locally, apparently `semver` adheres to "strict" semantic versioning, meaning that you cannot omit patch and minor versions if you want to parse the version string as "strictly" `semver::Version` (for good reasons).

i.e. this will be rejected, because `19` is not a valid semver (neither is `19.2`, but `19.0.0` is okay):

```
//@ min-llvm-version: 19
```

We *can* facade `semver`'s parsing logic and just stub out zeros, but it might be easier just to hand-write the parsing in the first place...

Reply:
> We can facade semver's parsing logic and just stub out zeros, but it might be easier just to hand-write the parsing in the first place... What do you think?

What about using zeros for the missing fields? Can't we treat "19" as "19.0.0"? This might be easier to handle and maintain.

Reply:
I can do that.

Reply:
Changed to use `semver` as suggested.